src/Microsoft.Management.Deployment/PackageManager.cpp

```diff
@@ -439,24 +417,38 @@ namespace winrt::Microsoft::Management::Deployment::implementation
                     if (installProgressOptional.has_value())
                     {
                         installProgress = installProgressOptional.value();
-                        ::SetEvent(progressEvent.get());
+                        progressEvent.SetEvent();
                     }
                     return;
                 }
             );
-            cancellationToken.callback([&queueItem]
+
+            std::weak_ptr<Execution::OrchestratorQueueItem> weakQueueItem(queueItem);
+            cancellationToken.callback([weakQueueItem, &canCancelQueueItem, &cancelWaitEvent]
                 {
-                    // The cancellation of the AsyncOperation on the client triggers Cancel which causes the Execute to end.
-                    Execution::ContextOrchestrator::Instance().CancelQueueItem(*queueItem);
+                    if (canCancelQueueItem)
+                    {
+                        auto strongQueueItem = weakQueueItem.lock();
+                        if (strongQueueItem) {
+                            // The cancellation of the AsyncOperation on the client triggers Cancel which causes the Execute to end.
+                            Execution::ContextOrchestrator::Instance().CancelQueueItem(*strongQueueItem);
+                        }
+                    }
+                    else
+                    {
+                        cancelWaitEvent.SetEvent();
+                    }
                 });
 
             // Wait for completion or progress events.
             // Waiting for both on the same thread ensures that progress is never reported after the async operation itself has completed.
             bool completionEventFired = false;
-            HANDLE operationEvents[2];
+            bool cancelWaitEventFired = false;
+            HANDLE operationEvents[3];
             operationEvents[0] = progressEvent.get();
             operationEvents[1] = queueItem->GetCompletedEvent().get();
-            while (!completionEventFired)
+            operationEvents[2] = cancelWaitEvent.get();
+            while (!completionEventFired && !cancelWaitEventFired)
             {
                 DWORD dwEvent = WaitForMultipleObjects(
                     2 /* number of events */,

```

Comment:
Maybe that should be `ARRAY_SIZE(operationEvents)`.

Reply:
:|